File: src/main/java/io/github/saneea/dvh/utils/EncodingRecognizer.java
package io.github.saneea.dvh.utils;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Map;

import io.github.saneea.dvh.utils.ByteSequenceRecognizer.ByteNode;

public class EncodingRecognizer {

	private static final Map<int[], Charset> BOM_TO_CHARSET = Map.of(//
			new int[] { 0xFE, 0xFF }, StandardCharsets.UTF_16BE, //
			new int[] { 0xFF, 0xFE }, StandardCharsets.UTF_16LE, //
			new int[] { 0xEF, 0xBB, 0xBF }, StandardCharsets.UTF_8 //
	);

	private static final ByteNode<Charset> BOM_TREE = ByteNode.fromMap(BOM_TO_CHARSET);

	public static ByteSequenceRecognizer<Charset> recognize(//
			InputStream inBinStream) throws IOException {
		return new ByteSequenceRecognizer<>(inBinStream, BOM_TREE);
	}

}

File: src/main/kotlin/io/github/saneea/dvh/utils/EncodingRecognizer.kt
package io.github.saneea.dvh.utils

import java.io.InputStream
import java.nio.charset.StandardCharsets

private val BOM_TREE = mapOf(
    intArrayOf(0xFE, 0xFF) to StandardCharsets.UTF_16BE,
    intArrayOf(0xFF, 0xFE) to StandardCharsets.UTF_16LE,
    intArrayOf(0xEF, 0xBB, 0xBF) to StandardCharsets.UTF_8
).byteNode


fun encodingRecognizer(inBinStream: InputStream) = ByteSequenceRecognizer(inBinStream, BOM_TREE)
